upstairs/src/lib.rs

```diff
@@ -1506,56 +1509,152 @@ impl Buffer {
         self.len == 0
     }
 
-    pub async fn as_vec(&self) -> MutexGuard<'_, Vec<u8>> {
-        self.data.lock().await
+    pub fn write(&mut self, offset: usize, data: &[u8]) {
+        assert!(offset + data.len() <= self.data.len());
+
+        self.data[offset..][..data.len()].copy_from_slice(data);
+        self.owned[offset..][..data.len()].fill(true);
+    }
+
+    pub fn write_with_ownership(
+        &mut self,
+        offset: usize,
+        data: &[u8],
+        owned: &[bool],
+    ) {
+        assert!(offset + data.len() <= self.data.len());
+        for i in 0..data.len() {
+            if owned[i] {
+                self.data[offset + i] = data[i];
+                self.owned[offset + i] = true;
+            }
+        }
```

Comment:
Use copy_from_slice and fill?

Reply:
I'm not sure I can with the conditional there, though https://github.com/oxidecomputer/crucible/pull/1091 will be able to per-block.